tests/h/services/bulk_executor/_actions_test.py

```diff
@@ -0,0 +1,112 @@
+
+import pytest
+from h_matchers import Any
+
+from h.h_api.bulk_api import Report
+from h.h_api.exceptions import ConflictingDataError, UnsupportedOperationError
+from h.models import User
+from h.services.bulk_executor._actions import UserUpsertAction
+from tests.h.services.bulk_executor.conftest import CommandFactory
+
+
+def assert_models_match_data(models_by_name, attrs_by_name, handlers=None):
+    assert set(models_by_name.keys()) == set(attrs_by_name.keys())
+
+    for name, model in models_by_name.items():
+        expected_attrs = attrs_by_name[name]
+
+        for field, value in expected_attrs.items():
+            found = getattr(model, field)
+
+            if handlers and field in handlers:
+                found = handlers[field](found)
+
+            assert found == value
+
+
+class TestBulkUserUpsert:
+    def test_it_can_insert_new_records(self, db_session, commands):
+        reports = UserUpsertAction(db_session).execute(commands)
+
+        assert reports == Any.iterable.comprised_of(Any.instance_of(Report)).of_size(3)
+
+        self.assert_users_match_commands(db_session, commands)
+
+    def test_it_can_update_records(self, db_session, commands):
+        update_commands = [
+            CommandFactory.user_upsert(i, extras={"display_name": f"changed_{i}"})
+            for i in range(3)
+        ]
+
+        action = UserUpsertAction(db_session)
+        action.execute(commands)
+        reports = action.execute(update_commands)
+
+        assert reports == Any.iterable.comprised_of(Any.instance_of(Report)).of_size(3)
+
+        self.assert_users_match_commands(db_session, update_commands)
+
+    def test_if_tails_with_duplicate_identities(self, db_session):
+        command_1 = CommandFactory.user_upsert(1)
+        command_2 = CommandFactory.user_upsert(2)
+        commands = [command_1, command_2]
+
+        command_2.body.attributes["identities"] = command_1.body.attributes[
+            "identities"
+        ]
+
+        with pytest.raises(ConflictingDataError):
+            UserUpsertAction(db_session).execute(commands)
+
+    def test_it_returns_in_the_same_order_as_the_commands(self, db_session, commands):
+        # Insert the values to set db order, then update them in reverse
+        action = UserUpsertAction(db_session)
+        reports = action.execute(commands)
+        reversed_reports = action.execute(list(reversed(deepcopy(commands))))
+
+        ids = [report.id for report in reports]
+        reversed_ids = [report.id for report in reversed_reports]
+
+        assert reversed_ids == list(reversed(ids))
+
+    @pytest.mark.parametrize("field", ["username", "authority"])
+    def test_it_fails_with_mismatched_queries(self, db_session, field):
+        command = CommandFactory.user_upsert(extras={field: "value"})
+        command.body.query[field] = "DIFFERENT"
+
+        with pytest.raises(UnsupportedOperationError):
+            UserUpsertAction(db_session).execute([command])
```

Comment:
You should test the exception's message using `match=`. This could be failing because of a disallowed query key rather than because of a query/body-attributes mismatch, and this test would still be passing.

Reply:
I think I'm going to stick with this for a few reasons:

 * The interface is that we raise this error when this condition arises, as long as we do, then this test should pass
 * The string of the error shouldn't be part of the interface
 * If we are concerned that this code doesn't explore all the code paths, then I think coverage has us... _covered_ (I'm here all night)

The last thing is to see the tests as boxing in the problem from different sides. The happy path sets a boundary proving we don't raise this exception every time. We then modify one thing which creates the situation we need to detect.

There are some things I think in this situation we can do:

 * Add a second parameter to the exception which allows them to be distinguished
 * Add more specific error classes to allow us to distinguish the cases by class

In this case, as the whole thing is really belt and braces (it should never happen with `h_api` mediating the communication) an assert would probably be enough, so I don't think it's justified.